src/python/pants/engine/legacy/parser.py

```diff
@@ -125,5 +128,32 @@ def parse(self, filepath, filecontent):
     # _intentional_ mutation would require a deep clone, which doesn't seem worth the cost at
     # this juncture.
     self._parse_context._storage.clear(os.path.dirname(filepath))
-    six.exec_(python, dict(self._symbols))
+    symbols = dict(self._symbols)
+
+    options, _ = OptionsInitializer(OptionsBootstrapper()).setup()
+    global_options = options.for_global_scope()
+    if global_options.build_file_imports != 'allow':
+      # This is not secure sandboxing, because people could replace __import__ again themselves
+      # if they wanted (and there are plenty of other ways to escape a python "sandbox"), but it
+      # should be sufficient to tell the casual user that they're doing something wrong.
+      builtins = dict(__builtins__)
+      if global_options.build_file_imports == 'warn':
+        import_hook = lambda import_name, *args: _warn_on_import(__import__, filepath, import_name, *args)
+      elif global_options.build_file_imports == 'error':
+        import_hook = _fail_on_import
+      else:
+        raise Exception("Didn't know what to do for build_file_imports value {}".format(global_options.build_file_imports))
+      builtins['__import__'] = import_hook
+      symbols['__builtins__'] = builtins
```

Comment:
any reason to not do this injection as part of `self._generate_symbols()` when `self._symbols` is computed?

Reply:
I need filepath to give a good error message in the warn case.